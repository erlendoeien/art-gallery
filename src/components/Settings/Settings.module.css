
.flexWrapper{
    display: flex;
    align-items: center;
}
.themeIcon{
    width: 25px; 
    margin-left: 20px; 
}

/* Toggle theme button  code: https://www.w3schools.com/howto/howto_css_switch.asp*/
.switch {
    position: relative;
    display: inline-block;
    width: 50px;
    height: 3vh;
}
.switch input {
    opacity: 0;
    width: 0;
    height: 0;
}
.slider {
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: #ee7752;
    -webkit-transition: 0.4s;
    transition: 0.4s;
    border-radius: 34px;
}
.slider:before {
    position: absolute;
    content: '';
    height: 3vh;
    width: 3vh;
    left: 0px;
    bottom: 0px;
    background-color: white;
    -webkit-transition: 0.4s;
    transition: 0.4s;
    border-radius: 50%;
}
input:checked + .slider {
    background-color: #23a6d5;
}
input:focus + .slider {
    box-shadow: 0 0 1px #23a6d5;
}
input:checked + .slider:before {
    -webkit-transform: translateX(26px);
    -ms-transform: translateX(26px);
    transform: translateX(26px);
}
