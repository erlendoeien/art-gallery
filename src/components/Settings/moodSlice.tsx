import { createSlice } from '@reduxjs/toolkit';
import { LocalStorage } from '../../utils/webStorageWrapper';

export enum Mood {
    moody = 'moody',
    happy = 'happy'
}

const storedIsMoody = LocalStorage.getItem('isMoody');
const initialState = {
    mood: storedIsMoody ? Mood.moody : Mood.happy
};

const moodSlice = createSlice({
    name: 'mood',
    initialState,
    reducers: {
        moodToggled(state) {
            if (state.mood === Mood.happy) {
                state.mood = Mood.moody;
            } else {
                state.mood = Mood.happy;
            }
        }
    }
});

export const { moodToggled } = moodSlice.actions;

export default moodSlice.reducer;
