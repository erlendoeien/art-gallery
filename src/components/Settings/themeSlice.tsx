import { createSlice } from '@reduxjs/toolkit';
import { LocalStorage } from '../../utils/webStorageWrapper';

export interface CustomTheme {
    fontColor: string;
    backgroundColor: string;
    navColor?: string;
    boxShadow?: string;
    isDarkTheme?: boolean;
}

const DARK_THEME: CustomTheme = {
    fontColor: '#ffffff',
    backgroundColor: '#1e1e1e',
    navColor: '#292828',
    boxShadow: '-1px 0px 15px 0px #000000'
    // isDarkTheme: true
};

const LIGHT_THEME: CustomTheme = {
    fontColor: '#1e1e1e',
    backgroundColor: '#ffffff',
    navColor: '#eeeeee',
    boxShadow: '-1px 0px 15px 0px #A6A6A6'
    // isDarkTheme: false
};

/**
 * Default context is light theme
 * TODO: Make it correct regarding initial async data
 */
export const initialState: CustomTheme = {
    ...(LocalStorage.getItem('isDarkTheme') === true ? DARK_THEME : LIGHT_THEME)
};

const themeSlice = createSlice({
    name: 'theme',
    initialState,
    reducers: {
        themeToggled(state) {
            state.isDarkTheme = !state.isDarkTheme;
            // Awful solution, but the state won't update updating the entire object
            if (!state.isDarkTheme) {
                state.backgroundColor = LIGHT_THEME.backgroundColor;
                state.boxShadow = LIGHT_THEME.boxShadow;
                state.fontColor = LIGHT_THEME.fontColor;
                state.navColor = LIGHT_THEME.navColor;
                // state = { ...LIGHT_THEME };
            } else {
                state.backgroundColor = DARK_THEME.backgroundColor;
                state.boxShadow = DARK_THEME.boxShadow;
                state.fontColor = DARK_THEME.fontColor;
                state.navColor = DARK_THEME.navColor;
                // state = { ...DARK_THEME };
            }
        }
    }
});

export const { themeToggled } = themeSlice.actions;

export default themeSlice.reducer;
