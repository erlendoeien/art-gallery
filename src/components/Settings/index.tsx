// eslint-disable-next-line
import React, { FunctionComponent, useContext } from 'react';
import styles from './Settings.module.css';
import { ThemeContext } from '../ThemeProvider';
import happy from '../../svg/emoji.png';
import sad from '../../svg/sad.png';
import sun from '../../svg/sun.png';
import moon from '../../svg/moon.png';
import { Mood, moodToggled } from './moodSlice';
import { useAppDispatch, useAppSelector } from '../../app/hooks';

const Settings: FunctionComponent = () => {
    const { fontColor, toggleTheme, isDarkTheme } = useContext(ThemeContext);
    const { mood } = useAppSelector((state) => state.mood);
    const dispatch = useAppDispatch();
    const isMoody = mood === Mood.moody;

    return (
        <div>
            <h2 style={{ color: fontColor }}>Tilpass siden</h2>
            <h3 style={{ color: fontColor }}>Velg mellom dark mode eller light mode:</h3>
            <div className={styles.flexWrapper}>
                <label htmlFor="darkmode" className={styles.switch}>
                    <input
                        id="darkmode"
                        type="checkbox"
                        onChange={toggleTheme}
                        checked={isDarkTheme}
                    />
                    <span className={styles.slider} />
                </label>
                <img
                    className={styles.themeIcon}
                    src={isDarkTheme ? moon : sun}
                    alt="Sol eller måne"
                />
            </div>

            <h3 style={{ color: fontColor }}>
                {' '}
                Hvilket humør er du i dag? <br />
                Velg mellom glade eller triste kunstverk:
            </h3>
            <div className={styles.flexWrapper}>
                <label htmlFor="mood" className={styles.switch}>
                    <input
                        id="mood"
                        type="checkbox"
                        onChange={() => dispatch(moodToggled())}
                        checked={isMoody}
                    />
                    <span className={styles.slider} />
                </label>
                <img
                    className={styles.themeIcon}
                    src={isMoody ? sad : happy}
                    alt="Glad eller Trist"
                />
            </div>
        </div>
    );
};

export default Settings;
