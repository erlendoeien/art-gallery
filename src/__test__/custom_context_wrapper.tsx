/* eslint react/prop-types: 0 */
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from '../components/ThemeProvider';
import { MoodProvider } from '../components/MoodProvider';

// Custom wrapper for render function,
// which wraps the component inside ThemeProvider and MoodProvider
export const Wrapper: React.ComponentType = ({ children }) => {
    return (
        <ThemeProvider>
            <MoodProvider>
                <MemoryRouter>{children}</MemoryRouter>
            </MoodProvider>
        </ThemeProvider>
    );
};

// re-export everything
export * from '@testing-library/react';
