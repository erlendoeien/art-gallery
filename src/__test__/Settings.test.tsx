import React from 'react';
import { render, Wrapper } from './custom_context_wrapper';
import Settings from '../components/Settings';

it('renders correctly', () => {
    const { container } = render(<Settings />, {
        wrapper: Wrapper
    });
    // snapshot testing
    expect(container).toMatchSnapshot();
});

test('renders title', () => {
    const { getByText } = render(<Settings />);
    expect(getByText('Tilpass siden')).toBeInTheDocument();
});

test('renders img alt text', () => {
    const { getByAltText } = render(<Settings />);
    expect(getByAltText('Glad eller Trist')).toBeInTheDocument();
});
