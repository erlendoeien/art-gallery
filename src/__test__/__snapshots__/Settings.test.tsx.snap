// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div>
    <h2
      style="color: rgb(30, 30, 30);"
    >
      Tilpass siden
    </h2>
    <h3
      style="color: rgb(30, 30, 30);"
    >
      Velg mellom dark mode eller light mode:
    </h3>
    <div
      class="flexWrapper"
    >
      <label
        class="switch"
        for="darkmode"
      >
        <input
          id="darkmode"
          type="checkbox"
        />
        <span
          class="slider"
        />
      </label>
      <img
        alt="Sol eller måne"
        class="themeIcon"
        src="sun.png"
      />
    </div>
    <h3
      style="color: rgb(30, 30, 30);"
    >
       
      Hvilket humør er du i dag? 
      <br />
      Velg mellom glade eller triste kunstverk:
    </h3>
    <div
      class="flexWrapper"
    >
      <label
        class="switch"
        for="mood"
      >
        <input
          checked=""
          id="mood"
          type="checkbox"
        />
        <span
          class="slider"
        />
      </label>
      <img
        alt="Glad eller Trist"
        class="themeIcon"
        src="sad.png"
      />
    </div>
  </div>
</div>
`;
